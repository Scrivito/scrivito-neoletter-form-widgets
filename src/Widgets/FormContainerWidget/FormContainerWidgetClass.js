import * as Scrivito from "scrivito";
const singleStepAllowedContent = [
  "AddressWidget",
  "BlogOverviewWidget",
  "BoxWidget",
  "ButtonWidget",
  "CarouselWidget",
  "ColumnContainerWidget",
  "DividerWidget",
  "EventOverviewWidget",
  "FactWidget",
  "FeaturePanelWidget",
  "FormCheckboxWidget",
  "FormDateWidget",
  "FormHiddenFieldWidget",
  "FormConditionalContainerWidget",
  "FormInputFieldWidget",
  "FormRatingWidget",
  "FormSelectWidget",
  "GalleryWidget",
  "GoogleMapsWidget",
  "GroupWidget",
  "HeadlineWidget",
  "IconContainerWidget",
  "IconWidget",
  "ImageWidget",
  "JobOverviewWidget",
  "LinkContainerWidget",
  "PricingWidget",
  "SpaceWidget",
  "TableWidget",
  "TestimonialSliderWidget",
  "TextWidget",
  "ThumbnailGalleryWidget",
  "TickListWidget",
  "VideoWidget",
  "VimeoVideoWidget",
  "YoutubeVideoWidget",
];
export const FormContainerWidget = Scrivito.provideWidgetClass(
  "FormContainerWidget",
  {
    attributes: {
      singleStepContent: ["widgetlist", { only: singleStepAllowedContent }],
      formId: "string",
      failedMessage: "string",
      submittedMessage: "string",
      submittingMessage: "string",
      hiddenFields: ["widgetlist", { only: "FormHiddenFieldWidget" }],
      formType: [
        "enum",
        {
          values: ["single-step", "multi-step"],
        },
      ],
      steps: ["widgetlist", { only: "FormStepWidget" }],
      forwardButtonText: "string",
      backwardButtonText: "string",
      submitButtonText: "string",
      showBorder: "boolean",
      singleSubmitButtonAlignment: [
        "enum",
        {
          values: ["left", "text-center", "text-end", "block"],
        },
      ],
    },
    extractTextAttributes: ["singleStepContent", "steps"],
  }
);
