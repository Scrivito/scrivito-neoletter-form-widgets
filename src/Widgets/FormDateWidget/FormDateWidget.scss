.scrivito-neoletter-form-widgets .form-date {
  .date-title {
    p {
      display: inline;
    }
  }
}
