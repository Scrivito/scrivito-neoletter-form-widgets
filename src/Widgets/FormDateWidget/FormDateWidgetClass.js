import * as Scrivito from "scrivito";

export const FormDateWidget = Scrivito.provideWidgetClass("FormDateWidget", {
  attributes: {
    title: "string",
    helpText: "html",
    required: "boolean",
    customFieldName: "string",
    dateType: [
      "enum",
      {
        values: ["date", "datetime-local"],
      },
    ],
  },
});
