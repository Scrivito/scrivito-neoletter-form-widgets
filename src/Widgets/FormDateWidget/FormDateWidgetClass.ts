import * as Scrivito from "scrivito";

export const FormDateWidget = Scrivito.provideWidgetClass("FormDateWidget", {
  attributes: {
    title: "html",
    helpText: "html",
    required: "boolean",
    customFieldName: "string",
    dateType: [
      "enum",
      {
        values: ["date", "datetime-local"]
      }
    ]
  }
});
