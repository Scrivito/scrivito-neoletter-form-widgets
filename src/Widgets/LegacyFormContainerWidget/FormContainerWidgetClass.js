import * as Scrivito from "scrivito";

export const FormContainerWidget = Scrivito.provideWidgetClass(
  "FormContainerWidget",
  {
    attributes: {
      content: "widgetlist",
      formId: "string",
      failedMessage: "string",
      submittedMessage: "string",
      submittingMessage: "string",
      hiddenFields: ["widgetlist", { only: "FormHiddenFieldWidget" }],
    },
    extractTextAttributes: ["content"],
  }
);
