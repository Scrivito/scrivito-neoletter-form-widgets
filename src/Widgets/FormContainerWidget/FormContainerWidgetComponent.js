import * as React from "react";
import * as Scrivito from "scrivito";
import { scrollIntoView } from "./utils/scrollIntoView";
import { FormFooterMultiSteps } from "./components/FormFooterMultiStepsComponent";
import { FormFooterSingleStep } from "./components/FormFooterSingleStepComponent";
import { FormHiddenFields } from "./components/FormHiddenFieldsComponent";
import "./FormContainerWidget.scss";
import { getScrivitoFormWidgetConfig } from "../../config/scrivitoConfig";

Scrivito.provideComponent("FormContainerWidget", ({ widget }) => {
  const tenant = getScrivitoFormWidgetConfig().tenant;
  if (!tenant) {
    return (
      <span className="missing-tenant">
        Warning! tenant has not been configured for the form widget.
      </span>
    );
  }
  const formEndpoint = `https://api.justrelate.com/neoletter/instances/${tenant}/form_submissions`;
  const [currentStep, setCurrentStepNumber] = React.useState(1);
  const [isSubmitting, setIsSubmitting] = React.useState(false);
  const [successfullySent, setSuccessfullySent] = React.useState(false);
  const [submissionFailed, setSubmissionFailed] = React.useState(false);
  const isSingleStep = widget.get("formType") == "single-step";
  const stepsLength = widget.get("steps").length;
  const isLastPage = currentStep == stepsLength;

  React.useEffect(() => {
    const steps = widget.get("steps");
    steps.forEach((step, i) => {
      const stepNumber = i + 1;
      step.update({
        stepNumber: stepNumber,
        isActive: currentStep == stepNumber,
      });
    });
  }, [widget.get("steps")]);

  if (isSubmitting) {
    return (
      <div className="form-container-widget text-center">
        <i className="fa fa-spin fa-spinner fa-2x" aria-hidden="true"></i>{" "}
        <span className="text-super">{widget.get("submittingMessage")}</span>
      </div>
    );
  }

  if (successfullySent) {
    return (
      <div className="form-container-widget text-center">
        <i className="fa fa-check fa-2x" aria-hidden="true"></i>{" "}
        <span className="text-super">{widget.get("submittedMessage")}</span>
      </div>
    );
  }

  if (submissionFailed) {
    return (
      <div className="form-container-widget text-center">
        <i className="fa fa-exclamation-triangle fa-2x" aria-hidden="true"></i>{" "}
        <span className="text-super">{widget.get("failedMessage")}</span>
      </div>
    );
  }

  return (
    <div
      className={`form-container-widget ${
        widget.get("showBorder") ? "form-border" : ""
      }`}
    >
      <form method="post" id={widget.get("formId")}>
        <FormHiddenFields widget={widget} />
        <Scrivito.ContentTag
          content={widget}
          attribute={isSingleStep ? "singleStepContent" : "steps"}
        />
      </form>
      {isSingleStep ? (
        <FormFooterSingleStep widget={widget} onSubmit={onSubmit} />
      ) : (
        <FormFooterMultiSteps
          widget={widget}
          onSubmit={onSubmit}
          onPageChange={onPageChange}
          currentStep={currentStep}
          stepsLength={stepsLength}
          isLastPage={isLastPage}
        />
      )}
    </div>
  );

  async function onSubmit() {
    if (Scrivito.isInPlaceEditingActive()) {
      return;
    }
    const isValid = validateCurrentStep();
    if (!isValid) {
      return;
    }
    const formElement = document.getElementById(widget.get("formId"));
    scrollIntoView(formElement);

    indicateProgress();
    try {
      await submit(formElement, formEndpoint);
      indicateSuccess();
    } catch (e) {
      setTimeout(() => {
        throw e;
      }, 0);

      indicateFailure();
    }
  }

  function indicateProgress() {
    setIsSubmitting(true);
    setSuccessfullySent(false);
    setSubmissionFailed(false);
  }

  function indicateSuccess() {
    setIsSubmitting(false);
    setSuccessfullySent(true);
    setSubmissionFailed(false);
  }

  function indicateFailure() {
    setIsSubmitting(false);
    setSuccessfullySent(false);
    setSubmissionFailed(true);
  }

  function validateCurrentStep() {
    return doValidate(widget.get("formId"), currentStep, isSingleStep);
  }

  function onPageChange(next) {
    let isValid = true;

    if (Scrivito.isInPlaceEditingActive()) {
      return;
    }
    if (next) {
      isValid = validateCurrentStep();
    }
    if (!isValid) {
      return;
    }
    const stepNumber = next
      ? Math.min(currentStep + 1, stepsLength)
      : Math.max(currentStep - 1, 1);
    setCurrentStepNumber(stepNumber);
    const formElement = document.getElementById(widget.get("formId"));
    scrollIntoView(formElement);
  }
});

async function submit(formElement, formEndpoint) {
  const data = new FormData(formElement);
  const dataToSend = new FormData();
  // workaround to send all field-names with equal name
  // as a comma separated string
  for (const [name, value] of data) {
    if (dataToSend.has(name)) {
      continue;
    } else {
      dataToSend.set(name, data.getAll(name).join(", "));
    }
  }
  const body = new URLSearchParams(dataToSend);
  // console.log("submitting", Object.fromEntries(body.entries()));
  const response = await fetch(formEndpoint, { method: "post", body });
  if (!response.ok) {
    throw new Error(
      `Response was not successful. Status code: ${response.status}.`
    );
  }
}

function doValidate(formId, currentStep, isSingleStep) {
  let isValid = true;
  const form = document.getElementById(formId);
  if (form) {
    const stepOrForm = isSingleStep
      ? form
      : form.querySelectorAll(`[data-step-number='${currentStep}']`);
    if (stepOrForm) {
      const allInputs = isSingleStep
        ? stepOrForm.querySelectorAll("input, select, textarea")
        : stepOrForm.item(0)?.querySelectorAll("input, select, textarea") || [];
      for (const node of allInputs.values()) {
        if (!node.checkValidity()) {
          node.reportValidity();
          return (isValid = false);
        }
      }
    }
    return isValid;
  }
}
