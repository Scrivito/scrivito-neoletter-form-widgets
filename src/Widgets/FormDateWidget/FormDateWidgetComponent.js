import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";
import { HelpText } from "../FormContainerWidget/components/HelpTextComponent";
import { Mandatory } from "../FormContainerWidget/components/MandatoryComponent";
import "./FormDateWidget.scss";

Scrivito.provideComponent("FormDateWidget", ({ widget }) => {
  const [value, setValue] = React.useState("");
  const onChangeValue = (e) => {
    const d = new Date(e.target.value);
    setValue(d.toISOString());
  };
  return (
    <div className="form-date mb-3">
      <div className="date-title">
        <span className="text-super">{widget.get("title")}</span>
        {widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>

      <input
        onChange={onChangeValue}
        className="datepicker"
        type={widget.get("dateType")}
        required={widget.get("required")}
      ></input>
      <input type="hidden" name={getFieldName(widget)} value={value}></input>
    </div>
  );
});
