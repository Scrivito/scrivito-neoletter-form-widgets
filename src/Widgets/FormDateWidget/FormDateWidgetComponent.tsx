import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { HelpText } from "../FormStepContainerWidget/components/HelpTextComponent";
import { Mandatory } from "../FormStepContainerWidget/components/MandatoryComponent";
import { FormDateWidget } from "./FormDateWidgetClass";
import "./FormDateWidget.scss";

Scrivito.provideComponent(FormDateWidget, ({ widget }) => {
  const [value, setValue] = React.useState("");
  const onChangeValue = (e: React.BaseSyntheticEvent) => {
    const d = new Date(e.target.value);
    setValue(d.toISOString());
  };
  return (
    <div className="form-date mb-3">
      <div className="date-title">
        <span className="text-super">{widget.get("title")}</span>
        {widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>

      <input
        onChange={onChangeValue}
        className="datepicker"
        type={widget.get("dateType")!}
        required={widget.get("required")}></input>
      <input
        type="hidden"
        className="show-in-review"
        name={getFieldName(widget)}
        value={value}></input>
    </div>
  );
});
