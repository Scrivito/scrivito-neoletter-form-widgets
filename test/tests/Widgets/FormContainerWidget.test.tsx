import * as Scrivito from "scrivito";
import { FormDateWidget } from "../../../src/Widgets/FormDateWidget/FormDateWidgetClass";
import { FormContainerWidget } from "../../../src/Widgets/LegacyFormContainerWidget/FormContainerWidgetClass";
import "../../../src/Widgets/FormDateWidget/FormDateWidgetComponent";
import "../../../src/Widgets/LegacyFormContainerWidget/FormContainerWidgetComponent";
import PageRenderer from "../../helpers/pageRenderer";

Scrivito.configure({ tenant: "inMemory" });
jest.mock(
  "../../../src/Widgets/FormStepContainerWidget/utils/lodashPolyfills",
  () => ({
    ...jest.requireActual(
      "../../../src/Widgets/FormStepContainerWidget/utils/lodashPolyfills"
    ),
    isEmpty: jest.fn(() => true)
  })
);
const pageRenderer = new PageRenderer();
const widgetProps = {
  content: [],
  formId: "123",
  failedMessage: "Failed",
  submittedMessage: "Done",
  submittingMessage: "Submitting",
  hiddenFields: []
};
describe("FormContainerWidget", () => {
  it("does not render if instanceId is empty", () => {
    pageRenderer.render({
      body: [new FormContainerWidget(widgetProps)]
    });
    const form = document.querySelector("form");
    expect(form).not.toBeInTheDocument();
    jest.resetAllMocks();
  });

  it("renders the form with content", () => {
    const content = [
      new FormDateWidget({
        customFieldName: "custom_date",
        title: "date",
        dateType: "date"
      })
    ];
    pageRenderer.render({
      body: [new FormContainerWidget({ ...widgetProps, content: content })]
    });

    const form = document.querySelector("form");
    const dateItem = document.querySelector('input[type="date"]');
    expect(form).toBeInTheDocument();
    expect(dateItem).toBeInTheDocument();
  });
  it("renders winnie-the-pooh", () => {
    pageRenderer.render({
      body: [new FormContainerWidget(widgetProps)]
    });

    const input = document.querySelector('input[name="fax"]');
    expect(input).toBeInTheDocument();
  });

  it("renders correctly", () => {
    const content = [
      new FormDateWidget({
        customFieldName: "custom_date",
        title: "date",
        dateType: "date"
      })
    ];
    const tree = pageRenderer.getAsJSON({
      body: [new FormContainerWidget({ ...widgetProps, content: content })]
    });
    expect(tree).toMatchSnapshot();
  });
});
