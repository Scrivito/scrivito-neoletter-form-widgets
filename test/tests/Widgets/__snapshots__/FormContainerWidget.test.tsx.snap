// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormContainerWidget renders correctly 1`] = `
<div>
  <div>
    <div
      className="scrivito-neoletter-form-widgets form-container-widget"
    >
      <form
        action="https://api.justrelate.com/neoletter/instances//form_submissions"
        method="post"
        onSubmit={[Function]}
      >
        <input
          name="form_id"
          type="hidden"
          value="123"
        />
        <input
          name="url"
          type="hidden"
        />
        <div />
        <div
          aria-hidden="true"
          className="winnie-the-pooh"
        >
          <input
            autoComplete="off"
            name="fax"
            role="presentation"
            tabIndex={-1}
          />
        </div>
        <div>
          <div>
            <div
              className="form-date mb-3"
            >
              <div
                className="date-title"
              >
                <span
                  className="text-super"
                >
                  date
                </span>
              </div>
              <input
                className="datepicker"
                onChange={[Function]}
                required={false}
                type="date"
              />
              <input
                className="show-in-review"
                name="custom_date"
                type="hidden"
                value=""
              />
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
